lib/std/os/windows.zig

```diff
@@ -3974,6 +3988,37 @@ pub const PROCESS_MEMORY_COUNTERS_EX = extern struct {
     PrivateUsage: SIZE_T,
 };
 
+pub const GetProcessMemoryInfoError = error{
+    AccessDenied,
+    InvalidHandle,
+    Unexpected,
+};
+
+pub fn GetProcessMemoryInfo(hProcess: HANDLE, out: *PROCESS_MEMORY_COUNTERS) GetProcessMemoryInfoError!void {
+    var vmc: VM_COUNTERS = undefined;
+    const rc = ntdll.NtQueryInformationProcess(hProcess, .ProcessVmCounters, &vmc, @sizeOf(VM_COUNTERS), null);
+    switch (rc) {
+        .SUCCESS => {
+            out.* = PROCESS_MEMORY_COUNTERS{
```

Comment:
Would it make sense to have this function return the `VM_COUNTERS` struct instead? If combined with swapping out the type used by `ChildProcess.ResourceUsageStatistics.rusage` with `VM_COUNTERS`, it seems like this code would become a lot more straightforward (no more copying things between similar structs, no `out` pointer needed).

Reply:
> Would it make sense to have this function return the `VM_COUNTERS` struct instead? If combined with swapping out the type used by `ChildProcess.ResourceUsageStatistics.rusage` with `VM_COUNTERS`, it seems like this code would become a lot more straightforward (no more copying things between similar structs, no `out` pointer needed).

It would make sense to use `VM_COUNTERS` instead. I wrote this as a direct replacement for [GetProcessMemoryInfo](https://learn.microsoft.com/en-us/windows/win32/api/psapi/nf-psapi-getprocessmemoryinfo) in case a user wants to obtain the same struct and avoid confusion. 

I suppose my question is:  Should Zig's rewrites of `kernel32` functions in `ntdll` use the same return types as `kernel32`? If not, I could definitely change that.

Reply:
My understanding is:
- Zig wants to rely on ntdll wherever possible: https://github.com/ziglang/zig/issues/1840
- Zig only cares about providing win32 bindings for things that it directly depends on/uses internally: https://github.com/ziglang/zig/issues/4426 (https://github.com/ziglang/zig/issues/4426#issuecomment-858820245)

Together, I think those make it acceptable to abandon `kernel32` APIs if they don't make as much sense as something lower level.

As an example, see something like [`windows.QueryObjectName`](https://github.com/ziglang/zig/blob/38ee46dda31101f349ea9eee762908f924fd384c/lib/std/os/windows.zig#L1027-L1054). I believe the equivalent higher level API would be [`GetFileInformationByHandleEx`](https://learn.microsoft.com/en-us/windows/win32/api/winbase/nf-winbase-getfileinformationbyhandleex), but using `NtQueryObject` means that Zig can provide a lower level, more custom tailored API that doesn't have a direct higher level equivalent.

Reply:
I can confirm that what @squeek502 described is accurate.

Reply:
Thanks for the clarification!